inc/cappuccino/UtMap.hpp

```diff
@@ -0,0 +1,468 @@
+    using TtlIterator = typename std::list<TtlElement>::iterator;
+    
+    struct KeyedElement {
+        /// The element's value.
+        ValueType m_value;
+        /// The iterator into the ttl data structure.
+        TtlIterator m_ttl_position;
+    };
+
+    struct TtlElement {
+        /// The point in time in  which this element's value expires.
+        std::chrono::steady_clock::time_point m_expire_time;
+        /// The iterator to the Key in the data structure.
+        KeyedIterator m_keyed_elements_position;
+    };
+
+     auto doInsertUpdate(
+        const KeyType& key,
+        ValueType&& value,
+        std::chrono::steady_clock::time_point expire_time,
+        Allow allow) -> bool;
+
+    auto doInsert(
+        const KeyType& key,
+        ValueType&& value,
+        std::chrono::steady_clock::time_point expire_time) -> void;
+
+    auto doUpdate(
+        KeyedIterator keyed_position,
+        ValueType&& value,
+        std::chrono::steady_clock::time_point expire_time) -> void;
+
+    auto doDelete(
+        KeyedIterator keyed_elements_position) -> void;
+
+    auto doFind(
+        const KeyType& key) -> std::optional<ValueType>;
+
+    auto doPrune(
+        std::chrono::steady_clock::time_point now) -> size_t;
+
+    /// Thread lock for all mutations.
+    Lock<SyncType> m_lock;
+
+    /// The keyed lookup data structure, the value is the KeyedElement struct which includes the value and an iterator to the associated m_ttl_list TTlElement.
+    typename std::map<KeyType, KeyedElement> m_keyed_elements;
+
+    /// The ttl sorted deque.
+    typename std::list<TtlElement> m_ttl_list;
+
+    /// The uniform TTL for every key value pair inserted into the cache.
+    std::chrono::milliseconds m_uniform_ttl;
+};
+} // namespace cappuccino
+
+namespace cappuccino
+{
+
+template <typename KeyType, typename ValueType, Sync SyncType>
+UtMap<KeyType, ValueType, SyncType>::UtMap(
+    std::chrono::milliseconds uniform_ttl)
+{
+    m_uniform_ttl = uniform_ttl;
+}
+
+template <typename KeyType, typename ValueType, Sync SyncType>
+auto UtMap<KeyType, ValueType, SyncType>::Insert(
+        const KeyType& key,
+        ValueType value,
+        Allow allow) -> bool
+{
+    std::lock_guard guard { m_lock };
+    const auto now = std::chrono::steady_clock::now();
+    const auto expire_time = now + m_uniform_ttl;
+
+    doPrune(now);
+
+    return doInsertUpdate(key, std::move(value), expire_time, allow);
+}
+
+template <typename KeyType, typename ValueType, Sync SyncType>
+template <typename RangeType>
+auto UtMap<KeyType, ValueType, SyncType>::InsertRange(
+        RangeType&& key_value_range,
+        Allow allow) -> size_t
+{
+    size_t inserted { 0 };
+    std::lock_guard guard { m_lock };
+    const auto now = std::chrono::steady_clock::now();
+    const auto expire_time = now + m_uniform_ttl;
+
+    doPrune(now);
+
+    for (auto [key, value] : key_value_range) {
```

Comment:
this should be auto& I think so its not an intermediate copy, please check every loop

Reply:
Ack

Reply:
Ack